sdk/src/crypto/asn1/mod.rs

```diff
@@ -10,9 +10,288 @@
 // cryptographic-message-syntax crate located at:
 // https://github.com/indygreg/cryptography-rs/tree/main/cryptographic-message-syntax/src/asn1
 
-// We can not incorporate the entire crate directly because other parts of the
-// crate contain dependencies on blocking calls in reqwest. Those calls are not
-// available in WASM environment.
+// We can not incorporate the entire crate directly because it relies on x509-certificate which
+// relies on ring.
+
+// Migrating fully from bcder to der will eliminate the need for much of this code.
+
+use std::io::Write;
+
+use bcder::{
+    decode::{Constructed, DecodeError, Source},
+    encode::{self, PrimitiveContent, Values},
+    Captured, Mode, Oid, Tag,
+};
+// Use der crate for time types to avoid custom date parsing
+use der::{Decode, Encode};
+
+// Common ASN.1 types shared across multiple RFCs
+
+/// Helper function to reconstruct DER encoding from primitive content
+/// (tag + length + content bytes)
+pub(crate) fn reconstruct_der_bytes(tag: u8, content: &[u8]) -> Vec<u8> {
+    let mut der_bytes = Vec::with_capacity(2 + content.len());
+    der_bytes.push(tag);
+    der_bytes.push(content.len() as u8); // length (assumes < 128)
+    der_bytes.extend_from_slice(content);
+    der_bytes
+}
+
+/// Helper function to extract string content from DER bytes
+/// DER format: tag (1 byte) + length (1 byte) + content
+pub(crate) fn extract_der_content(der_bytes: &[u8]) -> &str {
+    if der_bytes.len() >= 2 {
+        let length = der_bytes[1] as usize;
+        if length < 128 && der_bytes.len() >= 2 + length {
+            if let Ok(s) = std::str::from_utf8(&der_bytes[2..2 + length]) {
+                return s;
+            }
+        }
+    }
+    ""
+}
+
+/// Algorithm identifier for use with bcder
+#[derive(Clone, Debug)]
+pub struct AlgorithmIdentifier {
+    /// The algorithm OID
+    pub algorithm: Oid,
+}
+
+impl PartialEq for AlgorithmIdentifier {
+    fn eq(&self, other: &Self) -> bool {
+        self.algorithm == other.algorithm
+    }
+}
```

Comment:
I believe you can derive `PartialEq` on the struct since `Oid` implements it already.